FFmpegInterop/Reader.cpp

```diff
@@ -0,0 +1,55 @@
+//*****************************************************************************
+//
+//	Copyright 2015 Microsoft Corporation
+//
+//	Licensed under the Apache License, Version 2.0 (the "License");
+//	you may not use this file except in compliance with the License.
+//	You may obtain a copy of the License at
+//
+//	http ://www.apache.org/licenses/LICENSE-2.0
+//
+//	Unless required by applicable law or agreed to in writing, software
+//	distributed under the License is distributed on an "AS IS" BASIS,
+//	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//	See the License for the specific language governing permissions and
+//	limitations under the License.
+//
+//*****************************************************************************
+
+#include "pch.h"
+#include "Reader.h"
+#include "SampleProvider.h"
+
+using namespace std;
+
+namespace winrt::FFmpegInterop::implementation
+{
+	Reader::Reader(_In_ AVFormatContext* formatContext, _In_ const map<int, SampleProvider*>& streamMap) :
+		m_formatContext(formatContext),
+		m_streamIdMap(streamMap)
+	{
+
+	}
+
+	void Reader::ReadPacket()
+	{
+		AVPacket_ptr packet{ av_packet_alloc() };
```

Comment:
>AVPacket_ptr packet{ av_packet_alloc() }; [](start = 2, length = 41)

Performing a heap allocation each time a packet is read is not ideal for perf, but it looks like fixing this would require a number of changes since the QueuePacket() method on the sample provider takes in a unique_ptr, which prevents us from using stack space here (like the old implementation of FFmpegReader::ReadPacket) and it also prevents us from making the AVPacket poolable. We could potentially track this as a future perf improvement (e.g by creating a pooled version of AVPacket_ptr which QueuePacket accepts as an argument). There might also be no point in optimizing this if the underlying ffmpeg demuxer doesn't pool buffer allocations.

Reply:
We can revisit pooling AVPackets in the future. For now I added an optimization to set AVDISCARD_ALL on deselected streams so that we don't waste cycles allocating and reading packets that we're just going to immediately throw away.